### PR TITLE
Fix receipt broadcast to listening clients; add unit tests

### PR DESCRIPTION
**Pull request by jimthematrix**
Fixing the broadcast of transaction receipts to listening clients.

A client can listen for receipts on websocket by connecting to the `ws://<host>:<port>/ws` endpoint. It will be registered with the fabconnect server once it sends the following payload:

```
{"type":"listenreplies"}
```

### CODE REVIEW

File: internal/rest/receipt/receiptstore.go [hunk 1]
```diff
@@ -92,11 +83,17 @@ func NewReceiptStore(config *conf.RESTGatewayConf) ReceiptStore {
...
 
-func (r *receiptStore) Init() error {
-	return r.persistence.Init()
+func (r *receiptStore) Init(ws ws.WebSocketChannels, persistence api.ReceiptStorePersistence) error {
+	r.ws = ws
+	if persistence != nil {
+		r.persistence = persistence
+		return nil
+	} else {
+		return r.persistence.Init()
```
Comment (peterbroadhurst) [inline internal/rest/receipt/receiptstore.go:95]: Trying to understand this - maybe a comment would be useful?

It seems like the caller of this asserts that `r.persistence` is **not** `nil`, by passing `nil` in for the `persistence` parameter.

File: internal/rest/receipt/receiptstore.go [hunk 2]
```diff
@@ -174,7 +171,7 @@ func (r *receiptStore) writeReceipt(requestID string, receipt map[string]interfa
 		// Check if the reason is that there is a receipt already
 		existing, qErr := r.persistence.GetReceipt(requestID)
 		if qErr == nil && existing != nil {
-			log.Warnf("%s: exiting   receipt: %+v", requestID, *existing)
+			log.Warnf("%s: existing   receipt: %+v", requestID, *existing)
```
Comment (peterbroadhurst) [inline internal/rest/receipt/receiptstore.go:174]: Note a whitespace mismatch after the change

File: internal/rest/receipt/receiptstore.go [hunk 3]
```diff
@@ -201,12 +198,6 @@ func (r *receiptStore) GetReceipts(res http.ResponseWriter, req *http.Request, p
 		return
 	}
 
-	res.Header().Set("Content-Type", "application/json")
```
Comment (peterbroadhurst) [inline internal/rest/receipt/receiptstore.go:204]: Why is this no longer required?

Is this explicitly done in `marshalAndReply`?

**Review by peterbroadhurst — APPROVED** [review]
couple of minor comments to review, along with approval

File: internal/rest/receipt/receiptstore.go (hunk 1 above)
Comment (jimthematrix) [reply internal/rest/receipt/receiptstore.go:95]: added inline comments for clarifying these branches (only for testing purposes)

File: internal/rest/receipt/receiptstore.go (hunk 2 above)
Comment (jimthematrix) [reply internal/rest/receipt/receiptstore.go:174]: fixed

File: internal/rest/receipt/receiptstore.go (hunk 3 above)
Comment (jimthematrix) [reply internal/rest/receipt/receiptstore.go:204]: indeed, moved to `marshalAndReply` for both this and `GetReceipt`

**Comment by jimthematrix** [comment]
I was getting the following error when trying to push a commit for addressing the review comments. So will leave them to a later PR.

```
remote: error: GH006: Protected branch update failed for refs/heads/receipt-broadcast.
remote: error: Required status check "DCO" is expected.
To github.com:hyperledger/firefly-fabconnect.git
 ! [remote rejected] receipt-broadcast -> receipt-broadcast (protected branch hook declined)
error: failed to push some refs to 'github.com:hyperledger/firefly-fabconnect.git'
```

**Pull request merged**